betty/extension/webpack/__init__.py

```diff
@@ -0,0 +1,171 @@
+        return Path(__file__).parent / "assets"
+
+    @property
+    def public_css_paths(self) -> list[str]:
+        return [
+            self.app.static_url_generator.generate("css/vendor.css"),
+        ]
+
+    def new_context_vars(self) -> dict[str, Any]:
+        return {
+            "webpack_js_entrypoints": set(),
+        }
+
+    @property
+    def filters(self) -> dict[str, Callable[..., Any]]:
+        return FILTERS
+
+    @property
+    def _entrypoint_providers(
+        self,
+    ) -> Sequence[WebpackEntrypointProvider & Extension]:
+        return [
+            extension
+            for extension in self._app.extensions.flatten()
+            if isinstance(extension, WebpackEntrypointProvider)
+        ]
+
+    async def generate(self, job_context: GenerationContext) -> None:
+        build_directory_path = await self._ensure_build_directory(
+            job_context=job_context,
+        )
+        await self._copy_build_directory(
+            build_directory_path, self._app.project.configuration.www_directory_path
+        )
+
+    async def prebuild(self, job_context: Context) -> None:
+        async with TemporaryDirectory() as working_directory_path_str:
+            build_directory_path = await self._new_builder(
+                Path(working_directory_path_str),
+                job_context=job_context,
+            ).build()
+            await self._copy_build_directory(
+                build_directory_path,
+                PREBUILT_ASSETS_DIRECTORY_PATH,
+            )
+
+    def _new_builder(
+        self,
+        working_directory_path: Path,
+        *,
+        job_context: Context,
+    ) -> Builder:
+        return Builder(
+            working_directory_path,
+            self._entrypoint_providers,
+            self._app.project.configuration.debug,
+            self._app.binary_file_cache,
+            self._app.renderer,
+            job_context=job_context,
+            localizer=self._app.localizer,
+        )
+
+    async def _copy_build_directory(
+        self,
+        build_directory_path: Path,
+        destination_directory_path: Path,
+    ) -> None:
+        await to_thread(
+            copytree,
+            build_directory_path,
+            destination_directory_path,
+            dirs_exist_ok=True,
+        )
+
+    async def _ensure_build_directory(
+        self,
+        *,
+        job_context: Context,
+    ) -> Path:
+        builder = self._new_builder(
+            self._app.binary_file_cache.with_scope("webpack").path,
+            job_context=job_context,
+        )
+        try:
+            # (Re)build the assets if `npm` is available.
+            return await builder.build()
+        except NpmUnavailable:
+            # Use prebuilt assets if they exist.
+            if self._prebuilt_assets_requirement.is_met():
+                return PREBUILT_ASSETS_DIRECTORY_PATH
+            else:
+                # This must never happen, as the extension requirements are
```

Comment:
This MAY actually happen because the requirements are static and cannot check against configurable cache keys. That means we DO need a user-facing runtime error with a helpful message.